aws/resource_aws_launch_template.go

```diff
@@ -0,0 +1,1071 @@
+	conn := meta.(*AWSClient).ec2conn
+
+	var ltName string
+	if v, ok := d.GetOk("name"); ok {
+		ltName = v.(string)
+	} else if v, ok := d.GetOk("name_prefix"); ok {
+		ltName = resource.PrefixedUniqueId(v.(string))
+	} else {
+		ltName = resource.UniqueId()
+	}
+
+	launchTemplateData, err := buildLaunchTemplateData(d, meta)
+	if err != nil {
+		return err
+	}
+
+	launchTemplateOpts := &ec2.CreateLaunchTemplateInput{
+		ClientToken:        aws.String(resource.UniqueId()),
+		LaunchTemplateName: aws.String(ltName),
+		LaunchTemplateData: launchTemplateData,
+	}
+
+	resp, err := conn.CreateLaunchTemplate(launchTemplateOpts)
+	if err != nil {
+		return err
+	}
+
+	launchTemplate := resp.LaunchTemplate
+	d.SetId(*launchTemplate.LaunchTemplateId)
+
+	log.Printf("[DEBUG] Launch Template created: %q (version %d)",
+		*launchTemplate.LaunchTemplateId, *launchTemplate.LatestVersionNumber)
+
+	return resourceAwsLaunchTemplateUpdate(d, meta)
+}
+
+func resourceAwsLaunchTemplateRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).ec2conn
+
+	log.Printf("[DEBUG] Reading launch template %s", d.Id())
+
+	dlt, err := conn.DescribeLaunchTemplates(&ec2.DescribeLaunchTemplatesInput{
+		LaunchTemplateIds: []*string{aws.String(d.Id())},
+	})
+	if err != nil {
+		if isAWSErr(err, ec2.LaunchTemplateErrorCodeLaunchTemplateIdDoesNotExist, "") {
+			log.Printf("[WARN] launch template (%s) not found - removing from state", d.Id())
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("Error getting launch template: %s", err)
+	}
+	if len(dlt.LaunchTemplates) == 0 {
+		log.Printf("[WARN] launch template (%s) not found - removing from state", d.Id())
+		d.SetId("")
+		return nil
+	}
+	if *dlt.LaunchTemplates[0].LaunchTemplateId != d.Id() {
+		return fmt.Errorf("Unable to find launch template: %#v", dlt.LaunchTemplates)
+	}
+
+	log.Printf("[DEBUG] Found launch template %s", d.Id())
+
+	lt := dlt.LaunchTemplates[0]
+	d.Set("name", lt.LaunchTemplateName)
+	d.Set("latest_version", lt.LatestVersionNumber)
+	d.Set("default_version", lt.DefaultVersionNumber)
+	d.Set("tags", tagsToMap(lt.Tags))
+
+	version := strconv.Itoa(int(*lt.LatestVersionNumber))
+	dltv, err := conn.DescribeLaunchTemplateVersions(&ec2.DescribeLaunchTemplateVersionsInput{
+		LaunchTemplateId: aws.String(d.Id()),
+		Versions:         []*string{aws.String(version)},
+	})
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[DEBUG] Received launch template version %q (version %d)", d.Id(), *lt.LatestVersionNumber)
+
+	ltData := dltv.LaunchTemplateVersions[0].LaunchTemplateData
+
+	d.Set("disable_api_termination", ltData.DisableApiTermination)
+	d.Set("ebs_optimized", ltData.EbsOptimized)
+	d.Set("image_id", ltData.ImageId)
+	d.Set("instance_initiated_shutdown_behavior", ltData.InstanceInitiatedShutdownBehavior)
+	d.Set("instance_type", ltData.InstanceType)
+	d.Set("kernel_id", ltData.KernelId)
+	d.Set("key_name", ltData.KeyName)
+	d.Set("monitoring", ltData.Monitoring)
```

Comment:
This needs to be updated to err check (since its non-scalar now) as well as correctly handle the new nested structure:

```go
if err := d.Set("monitoring", getMonitoring(ltData.Monitoring)); err != nil {
  return err
}
```

Reply:
ahhh makes sense. Thank you. I'll make those changes now.